File: surf-api-bukkit/surf-api-bukkit-api/src/main/java/dev/slne/surf/surfapi/bukkit/api/nms/SurfBukkitNmsBridge.java
package dev.slne.surf.surfapi.bukkit.api.nms;

import dev.slne.surf.surfapi.bukkit.api.SurfBukkitApi;
import dev.slne.surf.surfapi.bukkit.api.nms.bridges.SurfBukkitNmsCommonBridge;
import dev.slne.surf.surfapi.bukkit.api.nms.bridges.SurfBukkitNmsNbtBridge;
import dev.slne.surf.surfapi.bukkit.api.nms.bridges.SurfBukkitNmsStatsBridge;
import dev.slne.surf.surfapi.bukkit.api.nms.bridges.packets.SurfBukkitNmsPacketBridges;
import dev.slne.surf.surfapi.bukkit.api.nms.listener.NmsClientboundPacketListener;
import dev.slne.surf.surfapi.bukkit.api.nms.listener.NmsServerboundPacketListener;
import javax.annotation.ParametersAreNonnullByDefault;
import org.jetbrains.annotations.ApiStatus.NonExtendable;
import org.jetbrains.annotations.NotNull;

@NonExtendable
@ParametersAreNonnullByDefault
public interface SurfBukkitNmsBridge {

  SurfBukkitNmsStatsBridge getStatsBridge();

  SurfBukkitNmsCommonBridge getCommonBridge();

  SurfBukkitNmsPacketBridges getPacketBridges();

  SurfBukkitNmsNbtBridge getNbtBridge();

  void registerServerboundPacketListener(NmsServerboundPacketListener<?> listener);

  void unregisterServerboundPacketListener(NmsServerboundPacketListener<?> listener);

  void registerClientboundPacketListener(NmsClientboundPacketListener<?> listener);

  void unregisterClientboundPacketListener(NmsClientboundPacketListener<?> listener);

  @NotNull
  static SurfBukkitNmsBridge get() {
    return SurfBukkitApi.get().getNmsBridge();
  }
}

File: surf-api-bukkit/surf-api-bukkit-api/src/main/kotlin/dev/slne/surf/surfapi/bukkit/api/nms/SurfBukkitNmsBridge.kt
package dev.slne.surf.surfapi.bukkit.api.nms

import dev.slne.surf.surfapi.bukkit.api.nms.listener.NmsClientboundPacketListener
import dev.slne.surf.surfapi.bukkit.api.nms.listener.NmsServerboundPacketListener
import dev.slne.surf.surfapi.core.api.util.requiredService

@NmsUseWithCaution
interface SurfBukkitNmsBridge {

    fun registerServerboundPacketListener(listener: NmsServerboundPacketListener<*>)
    fun unregisterServerboundPacketListener(listener: NmsServerboundPacketListener<*>)

    fun registerClientboundPacketListener(listener: NmsClientboundPacketListener<*>)
    fun unregisterClientboundPacketListener(listener: NmsClientboundPacketListener<*>)

    companion object : SurfBukkitNmsBridge by bridge {
        val instance get() = bridge
    }
}

@NmsUseWithCaution
val bridge = requiredService<SurfBukkitNmsBridge>()
